src/App.ts

```diff
@@ -6,84 +6,158 @@
-  logger.info("Stopping Proxy PagoPa Server...");
+  logger.info("Stopping Proxy PagoPA Server...");
   server.close();
 }
 
-// Set Restful WS routes
+/**
+ * Set RESTful WS endpoints
+ * @param {core.Express} app - Express server to set
+ * @param {Configuration} config - PagoPa proxy configuration
+ * @param {redis.RedisClient} redisClient - The redis client used to store information sent by PagoPA
+ * @param {PagamentiTelematiciPspNodoAsyncClient} pagoPAClient - PagoPa SOAP client to call verificaRPT and attivaRPT services
+ */
 function setRestfulRoutes(
   app: core.Express,
   config: Configuration,
-  verificaRPTPagoPaClient: pagoPaSoapClient.PagamentiTelematiciPspNodoAsyncClient,
-  attivaRPTPagoPaClient: pagoPaSoapClient.PagamentiTelematiciPspNodoAsyncClient
+  redisClient: redis.RedisClient,
+  pagoPAClient: pagoPASoapClient.PagamentiTelematiciPspNodoAsyncClient
 ): void {
+  const jsonParser = bodyParser.json();
   app.get(
-    config.CONTROLLER.ROUTES.RESTFUL.PAYMENTS_CHECK,
+    config.CONTROLLER.ROUTES.RESTFUL.PAYMENT_REQUESTS_GET,
     (req: express.Request, res: express.Response) => {
-      logger.info("Serving Payment Check Request (GET)...");
       toExpressHandler(
-        PaymentController.checkPaymentToPagoPa(
-          config.PAGOPA,
-          verificaRPTPagoPaClient
-        )
+        PaymentController.getPaymentInfo(config.PAGOPA, pagoPAClient)
       )(req, res);
     }
   );
   app.post(
-    config.CONTROLLER.ROUTES.RESTFUL.PAYMENTS_ACTIVATION,
+    config.CONTROLLER.ROUTES.RESTFUL.PAYMENT_ACTIVATIONS_POST,
+    jsonParser,
     (req: express.Request, res: express.Response) => {
-      logger.info("Serving Payment Activation Request (POST)...");
       toExpressHandler(
-        PaymentController.activatePaymentToPagoPa(
-          config.PAGOPA,
-          attivaRPTPagoPaClient
-        )
+        PaymentController.activatePayment(config.PAGOPA, pagoPAClient)
       )(req, res);
     }
   );
+  app.get(
+    config.CONTROLLER.ROUTES.RESTFUL.PAYMENT_ACTIVATIONS_GET,
+    (req: express.Request, res: express.Response) => {
+      toExpressHandler(PaymentController.getActivationStatus(redisClient))(
+        req,
+        res
+      );
+    }
+  );
+
+  // Endpoint for OpenAPI handler
+  app.get("/specs/api/v1/swagger.json", GetOpenapi(publicApiV1Specs));
+}
+
+/**
+ * Define and return a method to start a SOAP Server for PagoPA
+ * @param {redis.RedisClient} redisClient - The redis client used to store information sent by PagoPA
+ * @param {redisTimeout} redisTimeout - The timeout to use for information stored into redis
+ * @return {(app: core.Express) => soap.Server} A method to execute for start server listening
+ */
+function getSoapServer(
+  redisClient: redis.RedisClient,
+  redisTimeout: number
+): (app: core.Express) => soap.Server {
+  // Configuration for SOAP endpoints and callback handler
+  const service = {
+    FespCdService: {
+      FespCdPortType: {
+        cdInfoPagamento: async (
```

Comment:
1. declare the body of this function outside `getSoapServer`, it makes it easier to test and it's easier to read
1. this function can't both be `async` and accept a callback - I believe you need just a callback (soap doesn't want async functions as handlers)

Reply:
not true, an async function is just a function that returns a `Promise<>`, you can call it from any function, given that you handle the promise:

```ts
async function fa(): Promise<string> {
    return "hello from async";
}

function fb(): void {
    fa().then(console.log, console.log);
}

fb();
// prints "hello from async"
```

Reply:
instead, if you give to `soap` the `fa()` function that returns a `Promise`, `soap` will not handle the promise and any error will be ignored

Reply:
`soap` wants a handler that is a function with a callback parameter and ignore its return value - this handler should have the structure of the `fb()` above (+params)

Reply:
> this function can't both be async and accept a callback 

this is false as `fa.then(callback) === await fa; callback()` so the promise is handled either way and the callback gets called. fb() returns immediately anyway.

> soap wants a handler that is a function with a callback parameter and ignores its return value

true, but making the function async does not hurt neither prevent this to work. it's just an alternative way to express `fa.then()`.

Reply:
`fa.then(callback) !== await fa; callback()`

the former provides the result of the promise to the callback, the latter will discard the result of the promise...

now, [this](https://github.com/vpulim/node-soap/blob/master/lib/server.js#L446) is how the handler gets called, what happens when the handler returns a promise?

Reply:
that's why in the code we do use the result of the promise: `callback(iCdInfoPagamentoOutput);`

In your example the handler returns "void" (undefined) which causes handleResult() to be skipped. I don't know the semantic that we want to be implemented here in regards.

Reply:
let's see if this makes sense:

1. `soap` calls the handler [here](https://github.com/vpulim/node-soap/blob/master/lib/server.js#L446)
1. the handler returns a (yet to be resolved promise)
1. `result` is not `undefined` (it's a `Promise<>`) so [handlerResult is called](https://github.com/vpulim/node-soap/blob/master/lib/server.js#L448) (the Promise may still be waiting for `PaymentController.setActivationStatus()` to complete)
1. now, `handled` [gets set to true](https://github.com/vpulim/node-soap/blob/master/lib/server.js#L417) since this is the first time `handleResult` gets called for this request
1. `handleResult` will consider the promise as the result (see [this logic](https://github.com/vpulim/node-soap/blob/master/lib/server.js#L424)) (`error.Fault` and `result` are both `undefined`)
1. next, either [objectToRpcXML](https://github.com/vpulim/node-soap/blob/master/lib/server.js#L427) or [objectToDocumentXML](https://github.com/vpulim/node-soap/blob/master/lib/server.js#L431) will try to serialize the `Promise<>` to a body (see [here](https://github.com/vpulim/node-soap/blob/12ba988f5f4626417042eb5834436c25183f6891/lib/wsdl.js#L1609) and [here](https://github.com/vpulim/node-soap/blob/12ba988f5f4626417042eb5834436c25183f6891/lib/wsdl.js#L1578)
1. after a while (it's a remote request so it takes a lot of time) the `await` completes and `handleResult` gets called again, unfortunately this time `handled` is `true` so the result gets ignored 

am I missing something?

Reply:
1-6 are right (about 7 is not clear why handleResult should be called again if the method already returned).

Now the question is, have we to return "undefined" and so skip handleResult ? We cannot return anything which is related to the result of calling PaymentController.setActivationStatus if not in a Promise.

Reply:
About (7): `handleResult` is the `callback` that has been passed to the handler, remember that the promise will resolve long after the handler returns (it's an async processing), so:

* the first time `handleResult` gets called by `soap` when the handler returns the promise
* the second time `handleResult` gets called by the logic running inside the promise when the blocked `await` completes

Reply:
maybe I'm missing something but it looks like handleResult is an internal method of node-soap (not the callback we pass to the handler): https://github.com/vpulim/node-soap/blob/master/lib/server.js#L414 and it is called only if our method returns something != undefined [here](https://github.com/vpulim/node-soap/blob/master/lib/server.js#L447)

Reply:
The code should look somewhat like the following:

```ts
cdInfoPagamento: (
  iCdInfoPagamentoInput: IcdInfoPagamentoInput,
  callback: (iCdInfoPagamentoOutput: IcdInfoPagamentoOutput) => void
) => PaymentController.setActivationStatus(
  iCdInfoPagamentoInput,
  redisTimeout,
  redisClient
).then(
  callback, 
  err => callback({
    Fault: {
     // convert the error to a node-soap fault
     // see https://github.com/vpulim/node-soap#soap-fault
    }
}),
//...
```

Reply:
In your example cdInfoPagamento returns a Promise (so I see no difference to the actual code where it is async apart from the error handling with 'Fault')

Reply:
`method` is our handler (lookup happens [here](https://github.com/vpulim/node-soap/blob/master/lib/server.js#L409)

our handler gets called [here](https://github.com/vpulim/node-soap/blob/master/lib/server.js#L446) - 1st parameter is the request args, 2nd is the callback

the fact that `handleResult` gets called in two ways (passed as callback and called by node-soap with the result of handler) is for providing both sync and async handlers (see [here](https://github.com/vpulim/node-soap#soaplistenserver-path-services-wsdl---create-a-new-soap-server-that-listens-on-path-and-provides-services))

very confusing design imho

Reply:
you are right, I've added the curly braces to avoid returning the promise

Reply:
mm, maybe I'm still missing something but now cdInfoPagamento returns 'undefined' which means that handleResult won't be called because of the check [here](https://github.com/vpulim/node-soap/blob/master/lib/server.js#L447) and [this line](https://github.com/vpulim/node-soap/blob/master/lib/server.js#L433) won't be reached so our callback is never called.

Reply:
ok now I get it ! it must return undefined to avoid calling handlResult twice : )